### ISSUE TITLE
[xenial] Camera app image freeze

### ISSUE THREAD

**Issue by ernesst** [ubports/ubuntu-touch#611]
- Device: N5
- Channel: Xenial
- Build: 389

### Steps to reproduce
Just take a picture

### Expected behavior
no freeze

### Actual behavior
The image in the camera app freezes, blocking to take another picture.
Video more : as soon as i activate the video mode the image freeze, i cannot even start to record

### Logfiles and additional information
```

phablet@ubuntu-phablet:~$ tail -f .cache/upstart/application-legacy-camera-app-.log
file:///usr/share/camera-app//GalleryView.qml:70:5: QML OrientationHelper: All the parents must have an id.
State saving disabled for PhotogridView-photogridView:QQuickPinchArea-galleryViewPinch:_usr_share_camera-app__PhotogridView.qml:127:9:gridView, class OrientationHelper
file:///usr/share/camera-app//OptionsOverlay.qml:174:5: QML Information: Binding loop detected for property "height"
file:///usr/share/camera-app//OptionsOverlay.qml:174:5: QML Information: Binding loop detected for property "height"
[PERFORMANCE]: Last frame took 32 ms to render.
file:///usr/share/camera-app//ViewFinderExportConfirmation.qml:75:17: QML Item: Binding loop detected for property "height"
file:///usr/share/camera-app//ViewFinderExportConfirmation.qml:75:17: QML Item: Binding loop detected for property "height"
file:///usr/share/camera-app//ViewFinderExportConfirmation.qml:75:17: QML Item: Binding loop detected for property "height"
file:///usr/share/camera-app//OptionButton.qml:24:39: Unable to assign [undefined] to QString
file:///usr/share/camera-app//SlideshowView.qml:405: TypeError: Cannot read property 'actions' of null
	ext1: 1, ext2: 0 
	ext1: 0, ext2: 0 
	ext1: 0, ext2: 0 
qml: Picture saved as /home/phablet/Pictures/com.ubuntu.camera/image20180519_170459162.jpg
	ext1: 1, ext2: 0 
[PERFORMANCE]: Last frame took 36 ms to render.
[PERFORMANCE]: Last frame took 76 ms to render.
	ext1: 1, ext2: 0 
	ext1: 1, ext2: 0 
file:///usr/share/camera-app//OptionsOverlay.qml:174:5: QML Information: Binding loop detected for property "height"
file:///usr/share/camera-app//OptionsOverlay.qml:174:5: QML Information: Binding loop detected for property "height"
qml: Flipped
file:///usr/share/camera-app//OptionsOverlay.qml:174:5: QML Information: Binding loop detected for property "height"
file:///usr/share/camera-app//OptionsOverlay.qml:174:5: QML Information: Binding loop detected for property "height"
qml: Flipped
qml: optionsOverlayClose
[PERFORMANCE]: Last frame took 113 ms to render.
file:///usr/share/camera-app//OptionsOverlay.qml:174:5: QML Information: Binding loop detected for property "height"
file:///usr/share/camera-app//OptionsOverlay.qml:174:5: QML Information: Binding loop detected for property "height"
qml: Flipped
file:///usr/share/camera-app//OptionsOverlay.qml:174:5: QML Information: Binding loop detected for property "height"
file:///usr/share/camera-app//OptionsOverlay.qml:174:5: QML Information: Binding loop detected for property "height"
qml: Flipped
[PERFORMANCE]: Last frame took 36 ms to render.
Camera framerate boundaries are set by the backend
Camera framerate boundaries are set by the backend
QObject::connect: No such slot AalImageCaptureControl::onPreviewReady()
Camera framerate boundaries are set by the backend
Camera framerate boundaries are set by the backend
Already have a texture id and video sink, not creating a new one
	ext1: 0, ext2: 0 
	ext1: 0, ext2: 0 
qml: Picture saved as /home/phablet/Pictures/com.ubuntu.camera/image20180519_170638328.jpg
Camera framerate boundaries are set by the backend
Camera framerate boundaries are set by the backend
```

**Comment by fourloop2002** [comment]
Happens for me too.  If do a long swipe left the camera app image goes black (frozen picture disappears).  Reselect camera app and can tale another picture and the picture freezes.  The cycle repeats itself ;-)

**Comment by ernesst** [comment]
In the mean time when i'm creating a screenshot, the phone is slightly freezing.

**Comment by abkro** [comment]
I can confirm this running Xenial r416 on hammerhead.
1. The image freezes as described, preventing you from taking another picture
2. Long swipe towards left makes the image disappear (black screen). Sliding back towards the right so that the camera app is once more active (i.e. finger slides towards left first, then right without being lifted off the screen) reactivates the camera. A picture can once more be taken, but the freeze issue reproduces itself.
Other issues with the camera. 
3.Video recording not possible (as mentioned above).
4. When tapping the 'rotate' icon, bottom right, everything stalls and you have to wait 3 seconds before anything happens.

**Comment by RM250J** [comment]
Same here with r425 on hammerhead, take a picture and the screen freezes on that taken image, until I press the onscreen button to swap to the front camera. After a few seconds it will switch and a live image from the camera is active again, I can then switch back to the back camera and take another picture.

**Comment by tomoqv** [comment]
Probably not related, but equally critical as the camera app won't launch at all on the Pro 5. https://github.com/ubports/ubuntu-touch/issues/667

**Comment by mariogrip** [comment]
https://github.com/ubports/libhybris/pull/5 did not seem to have effect on this issue sadly :(

**Comment by cibersheep** [comment]
Similar behaviour on Hammerhead 2018-W31
- You can take first picture with back camera. After that preview image freezes.
- Reopen app, and choose video: Start recording video. Preview image is frozen. Video file is black but it has audio.
- Reopen app. Choose front camera. First image is green. Then preview, and preview freezes. Shooting takes a picture

**Comment by ernesst** [comment]
User experience wise, I'm not sure it's a good idea to move it from ota 4 to 5. 
As I'm daily using my camera at work, this ciurent bug is stopping me to use my 16.04 as daily driver ( I'm patient :) ). However std users recieving the ota with a camera regression will be disapointed.

**Comment by Flohack74** [comment]
@ernesst We know its not optimal, but we also seem to not find the bug. How long should we delay OTA-4 after me have delayed it for 2 months now? There are a lot of other fixes waiting to be deployed, plus we are running on an unsupported Ubuntu core version, and we are risking to get security issues etc.

**Comment by Flohack74** [comment]
This is why we said we want to shift it. Its not critical imho

**Comment by trickyut** [comment]
Hi,
issue appears on OPO as well
BR,
tricky

**Comment by ZikZakFR** [comment]
I have the same issue and what seems to be a related side-effect in another application: "InstantFX".
When a picture is selected, the next step won't show the actual picture for applying effects

**Comment by rolandwells** [comment]
same behavior here
device: nexus 5
build: aosp_hammerhead-userdebug 5.1.1 LMY48W 689 test-keys
rear-facing camera: you can take pictures, but the preview freezes after taking the first picture
front-facing camera: preview is frozen as soon as you switch to this camera, pictures can still be taken
in all cases taking pictures once the shutter icon is pressed takes way too long

**Comment by mariogrip** [comment]
This has been a incredible hard issue to fix, we have spend countless hours figuring out why this happens. We have come closer, but it will not be done for OTA-5, i'm extremely sorry about that and we hope to get it in asap! and since we are looking at quicker releases it wont be too long until next OTA.

**Comment by Arc676** [comment]
Congratulations on OTA5!

I've noticed that simply summoning and dismissing the app switcher can be enough to unfreeze the camera app.

**Comment by UniversalSuperBox** [comment]
It's interesting that a similar issue is reported on the Oneplus One, since that uses the same Android version as this.

It's even more interesting a similar issue is reported on the [Nexus 4](https://github.com/ubports/ubuntu-touch/issues/959) since that's using Android 4.4 with Canonical's port (to my knowledge).

As discussed on the previous Q&A, we've met no success with debugging this issue. I hope that @mariogrip can come in to give a more technical analysis, but both of us have spent many days banging our heads against it so far.

In short, we want to fix this as much as everyone wants it fixed and have put in a proportional amount of work to no benefit. Our next steps are:

* Release OTA-6 with this bug, putting it into the release notes
* Continue looking into the similar issues on other devices, since all of them share a common thread
* Investigate new, maintained options for Android camera compatibility like Sailfish's [gst-droid](https://github.com/sailfishos/gst-droid). These come with their own headaches (like integration into trust-store to ask if you'd like to allow an app to access your camera), but should be better in the long run.

**Comment by leftside2** [comment]
16.04 OTA-7 just now. Can't video recording with Nexus5 hammerhead. Couldn't either even before OTA-7.

**Comment by lduboeuf** [comment]
#23

**Comment by rolandwells** [comment]
Update, same issues persist on most recent build. Thanks to all working on this.

build: aosp_hammerhead-userdebug 5.1.1 LMY48W 879 test-keys
rear-facing camera: you can take pictures, but the preview freezes after taking the first picture, video recording prompts "Capture failed" error
front-facing camera: preview is frozen as soon as you switch to this camera, pictures can still be taken, video recording prompts "Capture failed" error

in all cases taking pictures once the shutter icon is pressed takes way too long

**Comment by rubencarneiro** [comment]
fixed now. https://github.com/ubports/android_device_lge_hammerhead/pull/18

**Comment by UniversalSuperBox** [comment]
Fixed by ubports/android_device_lge_hammerhead#18 for hammerhead. This issue is not tracking other devices, but similar changes may need to be made for them.

**Comment by ernesst** [comment]
Well done, thanks

**Comment by cibersheep** [comment]
Claap your haaaaaands and saaaay THAAAANK YOOOOOU!

**Comment by giiba** [comment]
It seems the liveview feed still freezes when:
- going to and from the capture settings menu at the bottom of the screen 
- or when switching between video/still modes
A quick half swipe on the edge of the screen and LV resumes.

If I stay out of settings and stick to one mode the camera works flawlessly. Also just a note: while the UT media player cannot playback the files the video plays fine on my desktop in VLC.

**Comment by giiba** [comment]
Nice to see this fix, thanks!

It seems the liveview feed still freezes when:
- going to and from the capture settings menu at the bottom of the screen
- or when switching between video/still modes
A quick half swipe on the edge of the screen and LV resumes.

If I stay out of settings and stick to one mode the camera works flawlessly. Also just a note: while the UT media player cannot playback the files the video plays fine on my desktop in VLC.

**Comment by advocatux** [comment]
Just for the record (OTA-9), I confirm this issue is fixed.
- Device: Nexus 5 (`hammerhead`)
- Channel: `rc`
- Build: Ubuntu 16.04 (`2019-W18`)

**Comment by giovanniiannotti** [comment]
OTA 9 fixed the issue for me too.

Nexus 5
Channel rc
2019-W18

**Comment by lduboeuf** [comment]
camera is buggy again on devel ( N5 )

**Comment by UniversalSuperBox** [comment]
We've had three people sitting together trying to fix this bug for a few hours, and at least we have a much better understanding of what's going on...

The first reference of this bug we found is in [android camera2 preview data stop callback in the middle of the call on StackOverflow](https://stackoverflow.com/questions/47707172/android-camera2-preview-data-stop-callback-in-the-middle-of-the-call). The user was trying to perform calls similar to us and ending up in the same situation, a stopped preview. The fact it was on the Nexus 5, in the same situation as us, and there wasn't so much as a theory in the comments was... concerning.

The bug is first officially referenced in https://github.com/ubports/android_device_lge_hammerhead/commit/6a34b0d90938f808b052db880301156debda9c90, which is again concerning. Instead of actually fixing the problem, it just reports to another service that the problem is occurring.

In the AOSP camera app, we find our red flag. [Line 34 in RepeatFailureDetector.java](https://github.com/amirzaidi/Camera2AS/blob/master/app/src/main/java/com/android/camera/one/v2/errorhandling/RepeatFailureDetector.java#L34) basically states that Google needed to do several hacks in sequence to get the camera to work properly again... The first is restarting the camera, the second is reporting to the user that the camera failed in the hope they'll restart their device.

Basically we're trying to fix the issue that Google's engineers gave up on.

That leaves us without many ways forward, actually. There's a lot of complexity leading to the detection and remediation of the issue, complexity that I don't think we can or should reimplement. The most likely fix, even though it's awful, is Marius' change which causes the camera to be restarted every time it does *anything*. Maybe we can use Google's code to make the restart faster, but it's looking less likely by the minute.

**Comment by dark-eye** [comment]
Since reactivating the app seems to solve the issue wee might be able to add a work around in the camera app itself... until this issue is solved.

**Comment by Flohack74** [comment]
So lets test tomorrow´s update ^^

**Comment by cibersheep** [comment]
N5 RC, camera works as expected

**Comment by j2g2rp** [comment]
Works fine on N5 on RC
